core/src/databases/database.rs

```diff
@@ -27,278 +28,137 @@ impl ToString for DatabaseType {
-    ) -> Self {
-        Database {
-            project: project.clone(),
-            created: created,
-            data_source_id: data_source_id.to_string(),
-            database_id: database_id.to_string(),
-            name: name.to_string(),
-            db_type: DatabaseType::LOCAL,
+pub async fn query_database(
+    tables: &Vec<DatabaseTable>,
+    store: Box<dyn Store + Sync + Send>,
+    query: &str,
+) -> Result<(Vec<DatabaseResult>, TableSchema)> {
+    let table_ids_hash = tables.iter().map(|t| t.unique_id()).sorted().join("/");
+    let database = store
+        .upsert_database(&table_ids_hash, HEARTBEAT_INTERVAL_MS)
+        .await?;
+
+    let time_query_start = utils::now();
+
+    let result_rows = match database.sqlite_worker() {
+        Some(sqlite_worker) => {
+            let result_rows = sqlite_worker
+                .execute_query(&table_ids_hash, tables, query)
+                .await?;
+            result_rows
         }
-    }
+        None => Err(anyhow!(
+            "No live SQLite worker found for database {}",
+            database.table_ids_hash
+        ))?,
+    };
+
+    utils::done(&format!(
+        "DSSTRUCTSTAT Finished executing user query on worker: duration={}ms",
+        utils::now() - time_query_start
+    ));
+
+    let infer_result_schema_start = utils::now();
+    let table_schema = TableSchema::from_rows(&result_rows)?;
+    utils::done(&format!(
+        "DSSTRUCTSTAT Finished inferring schema: duration={}ms",
+        utils::now() - infer_result_schema_start
+    ));
+
+    utils::done(&format!(
+        "DSSTRUCTSTAT Finished query database: duration={}ms",
+        utils::now() - time_query_start
+    ));
+
+    Ok((result_rows, table_schema))
+}
 
-    pub async fn get_tables(
-        &self,
-        store: Box<dyn Store + Sync + Send>,
-    ) -> Result<Vec<DatabaseTable>> {
-        match self.db_type {
-            DatabaseType::REMOTE => Err(anyhow!("Remote DB not implemented.")),
-            DatabaseType::LOCAL => {
-                let (_, tables, _) = store
-                    .list_databases_tables(
-                        &self.project,
-                        &self.data_source_id,
-                        &self.database_id,
-                        None,
-                    )
-                    .await?;
-
-                Ok(tables
-                    .into_iter()
-                    // Ignore empty tables.
-                    .filter_map(|t| match t.schema() {
-                        None => None,
-                        Some(_) => Some(t),
-                    })
-                    .collect::<Vec<_>>())
-            }
-        }
+pub async fn invalidate_database(db: Database, store: Box<dyn Store + Sync + Send>) -> Result<()> {
+    if let Some(worker) = db.sqlite_worker() {
+        worker.invalidate_database(db.unique_id()).await?;
+    } else {
+        // If the worker is not alive, we delete the database row in case the worker becomes alive again.
```

Comment:
Here there is an edge-case that is a bit annoyinbg

If the worker is "temporarily" dead and comes alive again. We delete the `database` row, so the next time that specific DB is queried, there will be 1 / NB_SQLITE_WORKERS chance that it goes to the resurrected worker. If that happens and the previous DB isn't expired yet, the user will be seeing stale data.

I think it's OK-ish to keep as-is for now as it is unlikely, but the right fix if that becomes an issue will be to add some additional logic in the heartbeat: if a heartbeat is resurrecting a worker, then we should call an "expire all" endpoint on the worker to make sure all existing DBs are invalidated (since we no longer have a guarantee that they are up-to-date).

WDYT @spolu  ? Should this be done now or fine to keep as-is for now ?